cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,310 @@
+            confusion_matrices = [confusion_matrices]
+        measure_qubits = cast(
+            Sequence[Sequence['cirq.Qid']],
+            [measure_qubits] if isinstance(measure_qubits[0], ops.Qid) else measure_qubits,
+        )
+        if len(confusion_matrices) != len(measure_qubits):
+            raise ValueError(
+                f"len(confusion_matrices): {len(confusion_matrices)} should be equal to "
+                f"len(measure_qubits): {len(measure_qubits)}"
+            )
+        for i, (cm, q) in enumerate(zip(confusion_matrices, measure_qubits)):
+            if cm.shape != (2 ** len(q),) * 2:
+                raise ValueError(
+                    f"Shape mismatch for confusion matrix {cm} at index {i} corresponding to {q}."
+                    f"Confusion Matrix shape {cm.shape} should match {(2 ** len(q),) * 2}"
+                )
+        self._confusion_matrices = list(confusion_matrices)
+        self._measure_qubits = [list(q) for q in measure_qubits]
+        self._qubits = sorted(set(q for ql in measure_qubits for q in ql))
+
+    @property
+    def confusion_matrices(self) -> List[np.ndarray]:
+        """List of confusion matrices corresponding to `measure_qubits` qubit pattern."""
+        return self._confusion_matrices
+
+    @property
+    def measure_qubits(self) -> List[List['cirq.Qid']]:
+        """Calibrated qubit pattern for which individual confusion matrices were computed."""
+        return self._measure_qubits
+
+    @property
+    def qubits(self) -> List['cirq.Qid']:
+        """Sorted list of all calibrated qubits."""
+        return self._qubits
+
+    def _get_vars(self, qubit_pattern: Optional[Sequence[Sequence['cirq.Qid']]] = None):
+        if qubit_pattern is None:
+            qubit_pattern = self.measure_qubits
+        abcd = "abcdefghijklmnopqrstuvwxyz"
+
+        def qubits_to_abcd(qs: Sequence['cirq.Qid']):
+            assert len(qs) <= len(abcd), "No. of qubits should be <= 26."
+            ret = ''.join(abcd[self.qubits.index(q)] for q in qs)
+            return ret + ret.upper()
+
+        return ','.join(qubits_to_abcd(qs) for qs in qubit_pattern)
+
+    @functools.lru_cache()
+    def _confusion_matrix(self, qubits: Tuple['cirq.Qid']) -> np.ndarray:
+        ret = np.einsum(
+            f'{self._get_vars()}->{self._get_vars([qubits])}',
+            *[
+                cm.reshape((2, 2) * len(qs))
+                for qs, cm in zip(self.measure_qubits, self.confusion_matrices)
+            ],
+        ).reshape((2 ** len(qubits),) * 2)
+        return ret / ret.sum(axis=1)
+
+    def confusion_matrix(self, qubits: Optional[Sequence['cirq.Qid']] = None) -> np.ndarray:
+        """Returns a single confusion matrix constructed for the given set of qubits.
+
+        The single `2 ** len(qubits) x 2 ** len(qubits)` confusion matrix is constructed
+        using the individual smaller `self.confusion_matrices` by applying necessary
```

Comment:
what happens if there's qubit-overlap in the smaller matrices

Reply:
assertion is probably good for the time being. you'd need to figure out how to average the values in the redundant matrices. you'd probably need to do some sort of geometric mean, which I guess is close to the observed behavior if you take the k'th root of the resulting values

Reply:
Combining overlapping confusion matrices is something you probably shouldn't do without explicitly supplying some sort of model for how you understand these matrices are arising (like a CTMP from [arXiv:2006.14044 [quant-ph]](https://arxiv.org/abs/2006.14044)). For disjoint matrices you're able to just assume a product distribution, but I don't think there's such an obvious default for the overlapping case.

Reply:
Thanks for the explanation. I've added a check to raise a value error if there are any overlapping qubits in the input.